File: mobile/src/core/colorKey.ts
import { type Uuid, uuidString } from './uuid';

/**
 * A color of the « Équipe » palette, for groups and avatars (docs/CONTRACTS-V2.md §1). Stored as its exact text:
 * `Coral` and `''` are refused by the server (`invalid_color`); where a color is optional, null means « automatic »
 * (`automaticColor`). Declared in palette order (color pickers).
 */
export const COLOR_KEYS = ['indigo', 'violet', 'blue', 'teal', 'green', 'amber', 'orange', 'coral', 'pink'] as const;
export type ColorKey = (typeof COLOR_KEYS)[number];

/** The automatic colors, indexed by `djb2(id) % 9`; this order keeps the hues of the v1 avatars. */
export const AUTOMATIC_COLOR_ORDER: readonly ColorKey[] = [
  'blue',
  'indigo',
  'violet',
  'pink',
  'orange',
  'teal',
  'green',
  'coral',
  'amber',
];

const MASK_64 = (BigInt(1) << BigInt(64)) - BigInt(1);

/**
 * The color of a group or a person whose color is automatic: `h = 5381; for byte in utf8(uppercase uuid string):
 * h = h * 33 + byte` on a wrapping 64-bit unsigned integer, then `AUTOMATIC_COLOR_ORDER[h % 9]`.
 */
export function automaticColor(id: Uuid): ColorKey {
  let hash = BigInt(5381);
  const thirtyThree = BigInt(33);
  // The uppercase uuid string is ASCII: its UTF-8 bytes are its char codes.
  for (const character of uuidString(id)) {
    hash = (hash * thirtyThree + BigInt(character.charCodeAt(0))) & MASK_64;
  }
  return AUTOMATIC_COLOR_ORDER[Number(hash % BigInt(AUTOMATIC_COLOR_ORDER.length))]!;
}

/** `color`, or the automatic color of `id` when it is null. */
export function resolvedColor(color: ColorKey | null | undefined, id: Uuid): ColorKey {
  return color ?? automaticColor(id);
}

/** True for a stored color text of the palette (exact case). */
export function isColorKey(value: unknown): value is ColorKey {
  return typeof value === 'string' && (COLOR_KEYS as readonly string[]).includes(value);
}

/** The color of a stored text; null for NULL (automatic) and for a text unknown to this client. */
export function storedColor(text: string | null | undefined): ColorKey | null {
  return isColorKey(text) ? text : null;
}

/** French name of the color, for the accessibility labels of the pickers. */
export function colorLabel(key: ColorKey): string {
  switch (key) {
    case 'indigo':
      return 'Indigo';
    case 'violet':
      return 'Violet';
    case 'blue':
      return 'Bleu';
    case 'teal':
      return 'Turquoise';
    case 'green':
      return 'Vert';
    case 'amber':
      return 'Ambre';
    case 'orange':
      return 'Orange';
    case 'coral':
      return 'Corail';
    case 'pink':
      return 'Rose';
  }
}

File: Packages/TeamTasksKit/Sources/TeamTasksCore/Models/ColorKey.swift
import Foundation

/// A color of the « Équipe » palette, for groups and avatars (docs/CONTRACTS-V2.md §1).
///
/// Stored as its raw value (SQL `text` with a check constraint), exact case: `Coral` and `""` are refused
/// (`invalid_color`). Where a color is optional, nil means « automatic »: see `automatic(for:)`. The hex values are
/// a client design matter. Cases are declared in palette order (color pickers).
public enum ColorKey: String, Sendable, Hashable, Codable, CaseIterable, Identifiable {
    case indigo
    case violet
    case blue
    case teal
    case green
    case amber
    case orange
    case coral
    case pink

    public var id: String { rawValue }

    /// The automatic colors, indexed by `djb2(id) % 9`. This order keeps the hues of the v1 avatars
    /// (blue, indigo, purple, pink, orange, teal, green, red, brown).
    public static let automaticOrder: [ColorKey] = [
        .blue, .indigo, .violet, .pink, .orange, .teal, .green, .coral, .amber,
    ]

    /// The color of a group or a person whose color is automatic (NULL): `automaticOrder[djb2(id) % 9]`, where djb2
    /// is `h = 5381; for byte in utf8(uppercase uuid string): h = h &* 33 &+ byte` on a wrapping `UInt64`
    /// (`UUID.uuidString` is uppercase on every platform).
    public static func automatic(for id: UUID) -> ColorKey {
        var hash: UInt64 = 5381
        for byte in id.uuidString.uppercased().utf8 {
            hash = hash &* 33 &+ UInt64(byte)
        }
        return automaticOrder[Int(hash % UInt64(automaticOrder.count))]
    }

    /// `color`, or the automatic color of `id` when `color` is nil.
    public static func resolved(_ color: ColorKey?, for id: UUID) -> ColorKey {
        color ?? automatic(for: id)
    }
}
